tests/infra/clients.py

```diff
@@ -449,6 +448,7 @@ def request(self, request):
             except requests.exceptions.ReadTimeout:
                 raise TimeoutError
             time.sleep(0.1)
+            self.connection_timeout -= 0.1
```

Comment:
I've been trying a few variations of this within my PR as well. We should do the same thing in the CurlClient, and ideally have this timeout for `signed_request` as well as `request`.

Reply:
Also I think we should decrement while handling `SSLError`, before we check the timeout. This way round, when we actually go negative we'll return `None` from this function, but I think we want to `raise CCFConnectionException`.

Reply:
How about:

```
self.connection_timeout -= 0.1
if self.connection_timeout < 0:
  self.connection_timeout = 0
```

and then the connection timeout check on :446 become equality?

Reply:
That still means we'll return `None`, as on the next loop the `while` will fail?

Reply:
We can raise if we fall through?